kernel/src/net/socket/ip/stream/mod.rs

```diff
@@ -719,17 +751,61 @@ impl SocketEventObserver for StreamSocket {
     }
 }
 
-impl Drop for StreamSocket {
-    fn drop(&mut self) {
-        let state = self.state.write().take();
+impl State {
+    /// Calls `f` to set raw socket option.
+    ///
+    /// Note that for listening socket, `f` is called on all backlog sockets in `Listen` State.
+    /// That is to say, `f` won't be called on backlog sockets in `SynReceived` or `Established` state.
+    fn set_raw_option(&mut self, set_option: impl Fn(&mut dyn RawTcpSetOption)) {
+        match self {
+            State::Init(init_stream) => init_stream.set_raw_option(set_option),
+            State::Connecting(connecting_stream) => connecting_stream.set_raw_option(set_option),
+            State::Connected(connected_stream) => connected_stream.set_raw_option(set_option),
+            State::Listen(listen_stream) => listen_stream.set_raw_option(set_option),
+        }
+    }
 
-        let iface_to_poll = match state {
+    fn iface_to_poll(&self) -> Option<Arc<Iface>> {
```

Comment:
I think we should rename `State::iface_to_poll` to `State::iface` and change its return value to `Option<&Arc<Iface>>` and let the caller decide if it wants to do the cloning.